core/admin/admin-notices.php

```diff
@@ -654,12 +703,66 @@ public function admin_notices() {
 		}
 	}
 
+	/**
+	 * Used to log campaigns for plugins
+	 */
+	public function maybe_log_campaign() {
+		// check for variables
+		if ( empty( $_GET['plg_campaign'] ) || empty( $_GET['plg_campaign_name'] ) ) {
+			return;
+		}
+		$allowed_plgs = [
+			'elementor_image_optimization_campaign',
+			'elementor_ea11y_campaign',
+			'elementor_site_mailer_campaign',
+		];
+
+		// whitelist campaign names
+		if ( ! in_array( $_GET['plg_campaign_name'], $allowed_plgs, true ) ) {
+			return;
+		}
+
+		// check nonce
+		if ( ! isset( $_GET['plg_campaign_nonce'] ) || ! wp_verify_nonce( sanitize_key( $_GET['plg_campaign_nonce'] ), sanitize_key( $_GET['plg_campaign_name'] ) ) ) {
+			return;
+		}
+
+		// check capabilities
+		if ( ! current_user_can( 'manage_options' ) ) {
+			return;
+		}
+
+		// check for empty values
+		if ( empty( $_GET['plg_source'] ) || empty( $_GET['plg_medium'] ) ) {
+			return;
+		}
+
+		$campaign_data = [
+			'source' => sanitize_key( $_GET['plg_source'] ),
+			'campaign' => sanitize_key( $_GET['plg_campaign'] ),
+			'medium' => sanitize_key( $_GET['plg_medium'] ),
+		];
+
+		set_transient( sanitize_key( $_GET['plg_campaign_name'] ), $campaign_data, 30 * DAY_IN_SECONDS );
+	}
+
+	public static function add_plg_campaign_data( $url, $campaign_data ) {
```

Comment:
Why is it public?

Reply:
so we can use it from external classes?

Reply:
When you need to use it external just change it, not now..